blockstorage: Return on fatal flush errors

sedited opened the pull request:
The goal of this PR is to establish that fatal blockstorage flush errors should be treated as errors at their call site.

Prior to this patch `FlushBlockFile` may have failed without returning in `Chainstate::FlushStateToDisk`, leading to a potential write from `WriteBlockIndexDB` that may refer to a block that is not fully flushed to disk yet. By returning if either `FlushUndoFile` or `FlushBlockFile` fail, we ensure that no further write operations take place that may lead to an inconsistent database when crashing. Add `[[nodiscard]]` annotations to them such that they are not ignored in future.

Functions that call either `FlushUndoFile` or `FlushBlockFile`, need to handle these extra abort cases properly. Since `Chainstate::FlushStateToDisk` already produces an abort error in case of `WriteBlockIndexDB` failing, no extra logic for functions calling `Chainstate::FlushStateToDisk` is required.

Besides `Chainstate::FlushStateToDisk`, `FlushBlockFile` is also called by `FindBlockPos`, while `FlushUndoFile` is only called by `FlushBlockFile` and `WriteUndoDataForBlock`. For both these cases, the flush error is not further bubbled up. Instead, the error is logged and a comment is provided why bubbling up an error would be less desirable in these cases.

---

This pull request is part of a larger effort towards improving the shutdown / abort / fatal error handling in validation code. It is a first step towards implementing proper fatal error return type enforcement similar as proposed by theuni in this pull request [comment](https://github.com/bitcoin/bitcoin/pull/27711#issuecomment-1563561502). For ease of review of these critical changes, a first step would be checking that `AbortNode` leads to early and error-conveying returns at its call site. Further work for enforcing returns when `AbortNode` is called is done in https://github.com/bitcoin/bitcoin/pull/27862.

ryanofsky commented on src/node/chainstate.cpp:
```diff
@@ -207,7 +207,9 @@ ChainstateLoadResult LoadChainstate(ChainstateManager& chainman, const CacheSize
-            AbortNode("Background chainstate cleanup failed unexpectedly.");
+            const auto error = _("Background chainstate cleanup failed unexpectedly.");
+            AbortNode(error.original);
+            return {ChainstateLoadStatus::INTERRUPTED, error};
```
In commit "chainstate: Return on ValidatedSnapshotCleanup fatal failure" (360fe5a47f75514621ef62bc4c906f2112d3308f)

I think this should return FAILURE not INTERRUPTED so caller can know that an error happened, not just an early shutdown request.

But I'd actually suggest dropping this commit from the PR. #27862 should cover this case, and I opened it earlier because I think AbortNode calls in assumeutxo code and in flush code seem different. There should be less of a rationale for ignoring errors in assumeutxo.

TheCharlatan replied:
Yup, I'll drop, make this PR about the flush errors only and continue discussion on your PR.

ryanofsky reviewed (commented):
I'm not sure about all the changes in this PR because it seems like some validation code is intentionally written to ignore flush errors, and now they are returning early and skipping other work. It also seems reasonable that some failed flushes like periodic flushes might not need to be fatal errors.

But adding return values and nodiscard attributes could at least make it clear if errors are being ignored intentionally or not.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/27866#pullrequestreview-1475602374

> It also seems reasonable that some failed flushes like periodic flushes might not need to be fatal errors.

This is a good point, and I did not think of this before. A flush may fail once due to some sporadic error, but may succeed again at a later point in time. I'm still not sure though if it is really correct to ignore the flush error. If we fail to flush, but succeed in writing the block index, and then crash, we might have indexed a block, that has not been synced to disk. It also feels weird that we are currently okay with the block files failing to flush, while the call to `CoinsTip().Flush()` returns early on failure. Can you comment on this rationale?

hebasto commented:
> A flush may fail once due to some sporadic error, but may succeed again at a later point in time. I'm still not sure though if it is really correct to ignore the flush error.

At least, such a flush error can be logged.

TheCharlatan commented on src/validation.cpp:
```diff
@@ -2506,7 +2506,9 @@ bool Chainstate::FlushStateToDisk(
                 // First make sure all block and undo data is flushed to disk.
-                m_blockman.FlushBlockFile();
+                if (!m_blockman.FlushBlockFile()) {
+                    return false;
```
There are many cases where these fatal error codes are eventually ignored as you go up the call stack. I would like to keep this PR focused on the question whether these blockstorage flush functions should at least indicate that they ran into an error. If you want to get a sense of just how many such "ignored" cases there are, I have my own set of patches that achieve something similar to Cory's "bubble up", but with your `Result<T, E>` type and introduced granularly for each fatal error call site here: https://github.com/TheCharlatan/bitcoin/pull/13.

TheCharlatan replied:
> Maybe your goal is to enhance libbitcoinkernel applications ability to handle and report errors, and my goal is to improve clarify of validation code and find to potential bugs in it.

I think that hits it on the nail. In any case I updated the PR description with much of the body of the commit message to make it clearer what I am trying to achieve here and why I'd consider it an improvement.

stickies-v reviewed (commented):
Concept ACK

The blockstorage changes seem pretty straightforward, the `validation.cpp` one I'll need to dive deeper into since it's got quite a few callsites.

TheCharlatan commented on src/node/blockstorage.h:
```diff
@@ -90,9 +90,9 @@ class BlockManager
+    [[nodiscard]] bool FlushBlockFile(bool fFinalize = false, bool finalize_undo = false);
+    [[nodiscard]] bool FlushUndoFile(int block_file, bool finalize = false);
+    [[nodiscard]] bool FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigned int nHeight, CChain& active_chain, uint64_t nTime, bool fKnown);
     bool FindUndoPos(BlockValidationState& state, int nFile, FlatFilePos& pos, unsigned int nAddSize);
```
I can't think of a good reason why this shouldn't be `[[nodiscard]]` too. However, as I said [here](https://github.com/bitcoin/bitcoin/pull/27866#discussion_r1246307365) I am trying to keep this PR's scope limited to just the flush functions and their immediate call sites. I added the `[[nodiscard]]` to `FindBlockPos`, because it reports on the flush error now too. `FindUndoPos` on the other hand is not immediately affected by the flush errors. There are many such examples where we ignore error codes upon return throughout the kernel code and I'd like to tackle them bit by bit.

ryanofsky reviewed (commented):
Noticed in IRC libbitcoinkernel updates:

> \<achow101> it looks like the current pr is #27866, although #27861 has also been getting review

IMO, this PR could be an improvement, but right now seems like a draft change. I think the main thing that's missing is a clear description of how it changes behavior, and why the behavior change is safe.

ryanofsky commented on src/node/blockstorage.cpp:
```diff
@@ -541,17 +543,21 @@ void BlockManager::FlushBlockFile(bool fFinalize, bool finalize_undo)
     FlatFilePos block_pos_old(m_last_blockfile, m_blockfile_info[m_last_blockfile].nSize);
     if (!BlockFileSeq().Flush(block_pos_old, fFinalize)) {
         m_opts.notifications.flushError("Flushing block file to disk failed. This is likely the result of an I/O error.");
+        return false;
```
In commit "blockstorage: Return on fatal flush errors" (d1c92b57a72b62ffa202f1d3d357b59befdc9c12)

This skips trying to flush the undo file if flushing the block file fails. It think it would be safer and more conservative to keep previous behavior of flushing both files in sequence.

You can still return false if either flush call fails, but I don't see a benefit in changing behavior here and potentially making code less robust, when it's not necessary just to add a return code.

I think it would also make review simpler to get rid of this unnecessary change, since the other changes in this commit are already pretty complex.

ryanofsky commented on src/node/blockstorage.cpp:
```diff
@@ -734,7 +742,9 @@ bool BlockManager::WriteUndoDataForBlock(const CBlockUndo& blockundo, BlockValid
         if (_pos.nFile < m_last_blockfile && static_cast<uint32_t>(block.nHeight) == m_blockfile_info[_pos.nFile].nHeightLast) {
-            FlushUndoFile(_pos.nFile, true);
+            if (!FlushUndoFile(_pos.nFile, true)) {
+                return false;
```
In commit "blockstorage: Return on fatal flush errors" (d1c92b57a72b62ffa202f1d3d357b59befdc9c12)

This change seems fragile to me, and it's unclear what the benefits are. I think it would be great to add a log print here, but if this this code is going to be changed to return false, I think that should happen in a separate followup commit, not combined with other changes here, and have a clear explanation and justification.

IIUC, this change introduces an inconsistency. For all block data, and for vast majority of undo data, the data is considered to be successfully written if the write call succeeeds regardless of whether syncing and trimming succeeds later. But now with this change, for some small fraction of undo block data, the data will be considered not written if the syncing or trimming fails. And the fraction of blocks this will be true for will be essentially random due to the undo flush heuristic (described in a new [comment](https://github.com/sdaftuar/bitcoin/blob/03dfe04075668aa3b99a5d6d68adfcecffca0593/src/node/blockstorage.h#L129-L139) from #27746)

I understand that your [goal](https://github.com/bitcoin/bitcoin/pull/27866#discussion_r1254840912) here is to have libbitcoinkernel functions return more complete error information. I understand that goal is easier to achieve by changing behavior of validation and block storage code and having return early in certain cases. But fundamentally it is not necessary to change behavior of any existing code to have it return more error information. Without returning early, you can always accumulate errors and just return the error code at the end.

If returning early makes sense, I think it should be justified in its own terms and done carefully in smaller commits.

ryanofsky commented on src/validation.cpp:
```diff
@@ -2505,7 +2505,9 @@ bool Chainstate::FlushStateToDisk(
                 // First make sure all block and undo data is flushed to disk.
-                m_blockman.FlushBlockFile();
+                if (!m_blockman.FlushBlockFile()) {
+                    return false;
```
In commit "blockstorage: Return on fatal flush errors" (d1c92b57a72b62ffa202f1d3d357b59befdc9c12)

This line appears to implement the major changing in behavior which is avoiding the "write from 
`WriteBlockIndexDB` that may refer to a block that is not fully flushed to disk yet."

I wonder what this failure looks like in practice, though. Could we write a test for it? I think it would at least make sense to have a log print that explains the problem that's happening.

There are also a *lot* of callers to `FlushStateToDisk` and I haven't looked at all of them to determine if they are handling this correctly. Can you describe what they should be doing, and do you know if they actually are doing it? The commit message says "no extra logic for functions calling `Chainstate::FlushStateToDisk` is required" mentioning an "abort error", but there is no abort error here, only a "return false" so I'm not sure how this is supposed to work.

dergoegge replied:
fwiw I think it is fine to change the behavior here but also don't mind if it's done in a follow-up. To me this just seems like a weird quirk given that we do return a fatal error for other flush operations.

I think, a failed flush opens the node up to be crashed in any case, e.g.:
https://github.com/bitcoin/bitcoin/blob/5aa67eb3655a0023f0cf115176fc8d5bac53cdcd/src/net_processing.cpp#L2217-L2219

Cleanly shutting down seems preferable.

ryanofsky reviewed (commented):
Code review d1c92b57a72b62ffa202f1d3d357b59befdc9c12. I think the extra error reporting here is great, and some of the changes in behavior seem good too. But I think there is too much happening in one commit. There are 3-4 separate changes in behavior that could be implemented in separate commits, and I think would be more clearly understood and evaluated that way.

hebasto commented:
Concept ACK.

sedited committed:
blockstorage: Mark FindBlockPos as nodiscard

A false return value indicates a fatal error (disk space being too low),
so make sure we always consume this error code.

This commit is part of an ongoing process for making the handling of
fatal errors more transparent and easier to understand.

sedited committed:
blockstorage: Return on fatal block file flush error

By returning an error code if `FlushBlockFile` fails, the caller now has
to explicitly handle block file flushing errors. Before this change
such errors were non-explicitly ignored without a clear rationale.

Prior to this patch `FlushBlockFile` may have failed silently in
`Chainstate::FlushStateToDisk`. Improve this with a log line. Also add a
TODO comment to flesh out whether returning early in the case of an
error is appropriate or not. Returning early might be appropriate to
prohibit `WriteBlockIndexDB` from writing a block index entry that does
not refer to a fully flushed block.

Besides `Chainstate::FlushStateToDisk`, `FlushBlockFile` is also called
by `FindBlockPos`. Don't change the abort behavior there, since we don't
want to fail the function if the flushing of already written blocks
fails. Instead, just document it.

sedited committed:
blockstorage: Return on fatal undo file flush error

By returning an error code if either `FlushUndoFile` or `FlushBlockFile`
fail, the caller now has to explicitly handle block undo file flushing
errors. Before this change such errors were non-explicitly ignored
without a clear rationale.

Besides the call to `FlushUndoFile` in `FlushBlockFile`, ignore its
return code at its call site in `WriteUndoDataForBlock`. There, a failed
flush of the undo data should not be indicative of a failed write.

Add [[nodiscard]] annotations to `FlushUndoFile` such that its return
value is not just ignored in the future.

TheCharlatan commented:
Thank you for the excellent review @ryanofsky.

Updated d1c92b57a72b62ffa202f1d3d357b59befdc9c12 -> f7072be104a8cf0516def7a4d9fca4791a8d2269 ([kernelReturnOnAbort_1](https://github.com/TheCharlatan/bitcoin/tree/kernelReturnOnAbort_1) -> [kernelReturnOnAbort_2](https://github.com/TheCharlatan/bitcoin/tree/kernelReturnOnAbort_2),  [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelReturnOnAbort_1..kernelReturnOnAbort_2))

* Split the changes up into three commits
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27866#discussion_r1256284494), ignoring `FlushBlockFile` failure in `FindBlockPos`. Also added a comment explaining why the return type is ignored and added the suggested log line.
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27866#discussion_r1256329428), always attempting `FlushUndoFile` in `FlushBlockFIle`, even when the block file flushing fails and return a success code only at the end of the function.
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27866#discussion_r1256350394), ignoring the `FlushUndoFile` failure in `WriteUndoDataForBlock` and instead added a comment and a log message in case of flush failure.

TheCharlatan commented:
Rebased 5720741c14e617ed338f64361b23f4e66ec99e07 -> 7721ab9139013d70ef0f058754fabc5aaeda2246 ([kernelReturnOnAbort_2](https://github.com/TheCharlatan/bitcoin/tree/kernelReturnOnAbort_2) -> [kernelReturnOnAbort_3](https://github.com/TheCharlatan/bitcoin/tree/kernelReturnOnAbort_3),  [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelReturnOnAbort_2..kernelReturnOnAbort_3))

* Fixed conflict with https://github.com/bitcoin/bitcoin/pull/27746

stickies-v reviewed (approved):
ACK 7721ab9139013d70ef0f058754fabc5aaeda2246

The changes in `blockstorage.cpp` and `blockstorage.h` are strict improvements, increasing visibility in flushing failures. I found the change in `validation.cpp` much more difficult to review, given the wide variety of ways in which `FlushStateToDisk()` is used. I've found that:
- it improves behaviour in some ways. For example, if `FlushBlockFile()` fails and the `WriteBlockIndexDB()` call still succeeds, we can end up in a situation where we expect to have a block, but don't actually have the data on disk. I found this quite easy to replicate by modifying `FlushBlockFile()` to a no-op, pkill bitcoind during IBD, and then call `getblock` on the latest synced block. This PR fixes that.
- I don't see how it can degrade behaviour:
    - logically, it doesn't make sense to flush the block index and chainstate if the block data may be missing
    - we already return early in `FlushStateToDisk()` for flushing failures, just not for `FlushBlockFile()`
    - I don't think this can affect consensus, e.g. the `BlockValidationState &state` out-parameter is unaffected; `AcceptBlock()` is unaffected
    - this code path is only reached when `fDoFullFlush || fPeriodicWrite`, which pretty much happens only every hour, when we shutdown, or when we exceed cache sizes.

(thanks @dergoegge for helping me understand the latter two points)

ryanofsky reviewed (changes requested):
Code review 7721ab9139013d70ef0f058754fabc5aaeda2246. Thanks for splitting up the commits and adding helpful log prints and comments. I like everything in this PR except the new `return false` after `FlushBlockFile()` fails in `Chainstate::FlushStateToDisk`. I think stickies-v makes the best possible case for it here: https://github.com/bitcoin/bitcoin/pull/27866#pullrequestreview-1570159243, but the test mentioned there of commenting out the block flush code entirely seems not very realistic, and none of the other cleanup in the PR requires this change in behavior, so I would really like to see it moved to another PR and justified on its own terms if it is a change worth making.

dergoegge reviewed (approved):
Code review ACK 7721ab9139013d70ef0f058754fabc5aaeda2246

TheCharlatan commented:
Updated 7721ab9139013d70ef0f058754fabc5aaeda2246 -> d8041d4e042957660827313951b18c8dd9a99a16 ([kernelReturnOnAbort_3](https://github.com/TheCharlatan/bitcoin/tree/kernelReturnOnAbort_3) -> [kernelReturnOnAbort_4](https://github.com/TheCharlatan/bitcoin/tree/kernelReturnOnAbort_4),  [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelReturnOnAbort_3..kernelReturnOnAbort_4))

* Removed early return in `FlushStateToDisk`, replaced it with a TODO comment and a log line.
* Addressed @jonatack's comment, added severity-based logging.
* Addressed @stickies-v's comment, splitting up long log line.

stickies-v commented:
re-ACK d8041d4

I was okay with the validation logic change that we had before, but as I don't have a lot of familiarity with the complexities of `FlushStateToDisk` I am definitely okay with ryanofsky's suggestion to keep that for a later, more focused improvement and making this pretty much a no-brainer PR.

naumenkogs commented:
Concept ACK

ryanofsky reviewed (approved):
Code review ACK d8041d4e042957660827313951b18c8dd9a99a16. Thanks for making suggested changes. This is now an easier PR to review since it is only adding new return codes, not changing existing ones, and no longer changing the way flush errors are handled other than by adding log prints.

If I could make one more suggestion, it would be to switch the order of the second and third commits so the `FlushBlockFile` return value doesn't get added with some initial behavior in the second commit (f0207e00303a1030eca795ede231e3c0d94df061) which is later modified in the third commit (d8041d4e042957660827313951b18c8dd9a99a16). If the commit order were switched, the `FlushBlockFile` return value could just be determined one way with no code churn.

It may be a good idea to follow up this PR with more changes that change error handling beyond adding log prints and return codes, but since FlushStateToDisk is called so many places, I think changes like these are harder to review, so it would be better for them to happen in a dedicated PR.

ryanofsky reviewed (approved):
Code review ACK d8041d4e042957660827313951b18c8dd9a99a16

There should be no change in behavior other than logging here. Just to remind my self why, the reason is that it mostly adds new return values and doesn't change existing ones. The only case where it does change a return value is in `FlushBlockFile` in the third commit. But there is only one call to `FlushBlockFile`, and if that fails it just triggers a log print.
